walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,13 +11,79 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
+ 
 package com.google.sps;
-
+ 
+import static com.google.common.collect.ImmutableList.toImmutableList;
+ 
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Streams;
+import java.util.stream.Collectors;
+import java.util.HashMap;
 import java.util.Collection;
-
+import java.util.Collections;
+import java.util.ArrayList;
+ 
 public final class FindMeetingQuery {
+  private int nextAvailableStart = TimeRange.START_OF_DAY;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+ 
+    // Cannot support events longer than 24 hours (1440 minutes)
+    if((request.getDuration()) > 1440) {
+      return availableTimes;
+    }
+ 
+    ImmutableSet<String> requiredAttendees = ImmutableSet.copyOf(request.getAttendees());
+ 
+    // Get time ranges from events that include at least one required attendee
+    ImmutableList<TimeRange> unavailableTimes = 
+        Streams.stream(events)
+            .filter(event -> !Collections.disjoint(requiredAttendees, event.getAttendees()))
+            .map(Event::getWhen)
+            .collect(toImmutableList());
+ 
+    // No unavailable time ranges means the entire day is available
+    if (unavailableTimes.isEmpty()) {
+      availableTimes.add(TimeRange.WHOLE_DAY);
+      return availableTimes;
+    }
+
+    availableTimes = 
+        Streams.stream(unavailableTimes)
+            .map(currentUnavailableTime -> {
+                HashMap<TimeRange, Integer> content = findAvailableTimeRange(currentUnavailableTime, nextAvailableStart, request);
+                TimeRange timeAvailable = content.entrySet().iterator().next().getKey();
+                nextAvailableStart = content.get(timeAvailable);
+                return timeAvailable;
+            })
+            .filter(currentUnavailableTime -> currentUnavailableTime!=null)
+            .collect(Collectors.toList());
+
+    // Add time leftover as the final available time range
+    if(nextAvailableStart < TimeRange.END_OF_DAY) {
+      availableTimes.add(TimeRange.fromStartEnd(nextAvailableStart, TimeRange.END_OF_DAY, true));
+    }
+ 
+    return availableTimes;
+  }
+
+  private static HashMap<TimeRange, Integer> findAvailableTimeRange(TimeRange currentUnavailableTime, int nextAvailableStart, MeetingRequest request) {
+    HashMap<TimeRange, Integer> content = new HashMap<TimeRange, Integer>();
+    TimeRange timeRangeAvailable = null;
```

Comment:
nit: no need to explicitly set to null

Reply:
Got it! I now only set null after going through the if statement